Refactor prep for only creating database for search/pulldata if needed

seadowg opened the pull request:
Work towards #5620
~~Blocked by getodk/javarosa#719~~
~~Blocked by https://github.com/getodk/collect/pull/5710~~
~~Blocked by #5781~~

This adds a `DynamicPreloadParseProcessor` and logic that uses its output to skip DB creation for forms that don't have dynamic preload functions (`search` or `pulldata`). Java Rosa doesn't currently provide us with a good way to detect these functions so this is just a stub for the moment (the database will still always be created for forms with external secondary instances).

For QA: there should be no changes to behaviour here. The biggest risk areas are forms using search/pulldata as that's where the majority of the refactoring has occurred.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Spike out not loading external data

seadowg committed:
Add failing test for existing zipped media support

This feature is pretty broken as it only works after
the first load.

seadowg committed:
Update failing test to specify fixed behaviour

seadowg committed:
Fix media file unzipping

seadowg committed:
Move dynamic preload code to renamed package

seadowg committed:
Create dedicate initializer for JavaRosa

seadowg committed:
Move dynamic preload code to package

seadowg committed:
Remove FormLoaderDependency from dynamic preload code

seadowg committed:
Extract external data creation code to use case

seadowg committed:
Make sure not loading external data is tested

seadowg committed:
Use Wrapper class

seadowg committed:
Fix extra externalization

seadowg committed:
Convert method object class to use case object

Like the approach being used in https://github.com/getodk/collect/pull/5521

seadowg committed:
Remove Collect.getIntsance() from new code

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.kt:
```diff
@@ -120,3 +191,19 @@ class ApplicationInitializer(
+    var containsSearchOrPullData = containsPullData
+        private set
+
+    override fun readExternal(`in`: DataInputStream?, pf: PrototypeFactory?) {
```
On upgrade, this will fail, form loading from the cached representation will throw an exception, and form loading from XML will be invoked?

seadowg replied:
I had missed this while dealing with your other comments. There is actually a problem here, but it's not to do with `readExternal`. Cached versions of the form never had `DynamicPreloadExtra` added to their `FormDef`, so it will never be read from the external representation. However, there is code in Collect now that assumes that a `FormDef` will have a `DynamicPreloadExtra` and we'll most likely see a crash/parse error there. I'll look into it.

seadowg replied:
Ok fixed that in https://github.com/getodk/collect/pull/5665/commits/2ca6909068a9ac6dd24279add7f394861e34c147. 

I was then very confused as to why we didn't run into similar problems with entities as it looks like we also assume `FormDef` instances have a `EntityFormExtra` in `EntityFormFinalizationProcessor`. My guess is that because we added the concept of "extras" for `FormDef` in that release, cached `FormDef` instances would fail to deserialize as they wouldn't have the `ExternalizableExtras` included as part of them and then be deleted (silently). There's definitely a scary realisation in there for me that we don't have any versioning built (like we get from something like `Serializable`), so it's pretty easy to miss potential upgrade problems like this (and manually testing they don't exist is subtle). I'll add this as a conversation topic.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/dynamicpreload/DynamicPreloadXFormParserFactory.kt:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    override fun processBindAttribute(name: String, value: String, binding: DataBinding) {
+        val bindingCalculate = binding.calculate
```
We also need to look in `readonly`, `required`, `relevant`, `constraint` (all the [bind attributes](https://getodk.github.io/xforms-spec/#bind-attributes) that can be set to an XPath expression). I think that's it. You could look at how triggerables are identified to make sure that's exhaustive.

The `containsFunc` method checks in predicates, too, right?

lognaturel reviewed (changes requested):
One issue I thought of about where `pulldata` can show up. Looks good otherwise!

seadowg committed:
Add test for DynamicPreloadParseProcessor

seadowg committed:
Detect pulldata in readonly

seadowg committed:
Detect pulldata in required

seadowg committed:
Detect pulldata in relevant

seadowg committed:
Detect pulldata in constraint

seadowg commented:
@lognaturel I decided against adding a `FunctionProcessor` to JavaRosa as it seemed like it would be more work once I investigated it. Instead, the existing `DynamicPreloadParseProcessor` will now check all the bind attributes where `pulldata` might be present (the ones you suggested). As far as I can see, these are the only places as it's the only things that result in a `FormDef#addTriggerable` call.

seadowg committed:
Account for absolutes

seadowg commented:
@lognaturel I'd missed that there were conflicts on this sorry! Ready to go again.

seadowg commented:
@lognaturel scratch that! Looks like we have some failing instrumentation tests.

seadowg commented:
@lognaturel ...and scratch that as well. Tests are failing due to unrelated problem around updating JavaRosa (https://github.com/getodk/collect/pull/5710).

seadowg committed:
Allow cached FormDefs to still work

seadowg commented:
@lognaturel does it make sense to transfer this over to @grzesiek2010 for review?

lognaturel commented:
We should probably both take a look since we have different lenses. My review could come after an initial merge and I could file an issue if I think of follow-up work.

lognaturel commented:
Unfortunately, it just occurred to me that `pulldata` could show up in the `value` attribute of a `setvalue` action: https://getodk.github.io/xforms-spec/#events-and-actions This would be for a [dynamic default](https://docs.getodk.org/form-logic/#dynamic-defaults), for example. We really should have a way of authoritatively saying where function calls can occur but it seems we don't.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/application/Collect.java:
```diff
@@ -26,7 +26,7 @@
 import org.jetbrains.annotations.NotNull;
 import org.odk.collect.android.BuildConfig;
-import org.odk.collect.android.externaldata.ExternalDataManager;
+import org.odk.collect.android.dynamicpreload.ExternalDataManager;
```
`ExternalDataManager` should not be stored statically here.

seadowg committed:
Don't disable dynamic preload data for now

We've realized that pulldata can appear in more places so need a
better way to detect it than inspecting each location. We'll most
likely need to add a processor that lets us inspect XPathFuncExpr
instances as they are created during parsing.

seadowg commented:
@lognaturel there's a lot of good changes here and I don't want to hold it up anymore while we iterate on the best way to detect `search`/`pulldata`. I think we should just merge this as good prep.

lognaturel reviewed (approved):
As I went through this I started questioning the `ParseProcessor` approach to identifying search/pulldata from JR since their implementations live in Collect. That said, maybe it will still make sense to be able to identify a call on any function and I think most of the restructuring has value.

seadowg commented:
> That said, maybe it will still make sense to be able to identify a call on any function and I think most of the restructuring has value.

That's definitely what I'm thinking. It may be that the `ParseProcessor` gets replaced here with a hypothetical `XPathFuncExprProcessor` that allows Collect to inspect functions when they're being parsed. Another alternative might be to have JavaRosa expose all the unique functions used within a form. I'm also wondering if we can stick `containsFunc` at the `FormDef` level and have it traverse the form. I need to investigate if relevance would be an issue at that stage or not though.

srujner commented:
**Tested with Success!**

Verified on device with Android 13

Verified cases:
- Forms using search/pulldata;
- Regression check on big forms:
- Acceptance Criteria has been met:
- Rotate the screen, minimize the app

dbemke commented:
**Tested with Success!**

Verified on device with Android 10
